crates/environ/src/fact/trampoline.rs

```diff
@@ -456,12 +881,21 @@ impl Compiler<'_, '_> {
         let lift_opts = &adapter.lift.options;
         let lower_opts = &adapter.lower.options;
 
-        let src_flat =
-            self.types
-                .flatten_types(lift_opts, MAX_FLAT_RESULTS, src_tys.iter().copied());
-        let dst_flat =
+        let src_flat = self.types.flatten_types(
+            lift_opts,
+            if lift_opts.async_ {
+                MAX_FLAT_PARAMS
+            } else {
+                MAX_FLAT_RESULTS
+            },
+            src_tys.iter().copied(),
+        );
```

Comment:
This looks pretty similar to some code in `signature.rs` (and I'm similarly a bit confused about params/results a little). Is this similar enough that it might warrant a helper of some kind to do this internally?

(similar for `dst_flat` below, some sort of helper that takes `lower_opts` or `async_` which handles async-vs-not)